### PR TITLE
RBAC implementation for key value pair

### PR DESCRIPTION
**Pull request by shivani-orch**
Implemented RBAC functionality and unit tests for key-value pairs for existing and new permission types. Previously, RBAC feature for key value pairs are not yet implemented.

- Existing permission types - KEY_VALUE_VIEW, KEY_VALUE_SET, KEY_VALUE_DELETE
- New permission types - KEY_VALUE_LIST, KEY_VALUE_ALL

RBAC is enabled in the st2.conf file. Access to a key value pair is checked in the KeyValuePair API controller.

- SET and DELETE permissions will automatically grant LIST and VIEW permissions.
- By default, user has access to his/her own user scoped KVPs without requiring specific permission grant.
- A non-admin user can be explicitly granted permission to one or more system scoped KVPs.
- A non-admin user cannot access another user's KVPs.
- By default, admin and system_admin has ALL access to system scoped KVPs.
- Admin has full access to another user's KVPs (behavior in current version).

This change requires KeyValuePair API changes at https://github.com/StackStorm/st2/pull/5354.

### CODE REVIEW

**Comment by shivani-orch** [comment]
We couldn't resolve why circleci checks have failed. 
Following is the error - . virtualenv/bin/activate; flake8 --config=lint-configs/python/.flake8 st2rbac_backend/ tests/
/bin/sh: 85: virtualenv/bin/activate: /tmp/st2/st2api: Permission denied

File: st2rbac_backend/resolvers.py [hunk 1]
```diff
@@ -636,14 +721,55 @@ class KeyValuePermissionsResolver(PermissionsResolver):
     """
 
     resource_type = ResourceType.KEY_VALUE_PAIR
+    view_grant_permission_types = [
+        PermissionType.KEY_VALUE_VIEW,
+        PermissionType.KEY_VALUE_SET,
+        PermissionType.KEY_VALUE_DELETE,
+        PermissionType.KEY_VALUE_LIST,
+        PermissionType.KEY_VALUE_ALL,
+    ]
```
Comment (m4dcoder) [inline st2rbac_backend/resolvers.py:730]: Do we need to override user_has_resource_api_permission, abstract method for PermissionResolver at  https://github.com/StackStorm/st2-rbac-backend/blob/master/st2rbac_backend/resolvers.py#L84?

File: st2rbac_backend/resolvers.py [hunk 2]
```diff
@@ -636,14 +721,55 @@ class KeyValuePermissionsResolver(PermissionsResolver):
...
+        )
+
+        if has_system_role_permission:
+            self._log("Found a matching grant via system role", extra=log_context)
+            return True
+
+        # Check custom roles and permission grants
+        resource_uid = resource_db.get_uid()
+        resource_types = [ResourceType.KEY_VALUE_PAIR]
+        permission_types = [PermissionType.KEY_VALUE_ALL, permission_type]
```
Comment (m4dcoder) [inline st2rbac_backend/resolvers.py:759]: How come we are granting `PermissionType.KEY_VALUE_ALL` here by default?

File: tests/unit/test_rbac_resolvers_key_value_pair.py [hunk 3]
```diff
@@ -17,39 +17,310 @@
...
+        kvp_4_db = KeyValuePair.add_or_update(kvp_4_db)
+        self.resources["user_role_4"] = kvp_4_db
+
+        kvp_5_db = KeyValuePairDB(name="key5", value="val5")
+        kvp_5_db = KeyValuePair.add_or_update(kvp_5_db)
+        self.resources["user_role_5"] = kvp_5_db
+
+        kvp_6_db = KeyValuePairDB(name="key6", value="val6")
+        kvp_6_db = KeyValuePair.add_or_update(kvp_6_db)
+        self.resources["user_role_6"] = kvp_6_db
```
Comment (m4dcoder) [inline tests/unit/test_rbac_resolvers_key_value_pair.py:83]: Are these unit tests for system scope or user scope key value pairs? Do we have unit tests for both? It is not obvious here.

**Review by m4dcoder — CHANGES_REQUESTED** [review]

**Comment by m4dcoder** [comment]
Please make sure CIs are green before asking for reviews.

**Comment by m4dcoder** [comment]
@ashwini-orchestral The CIs are failing with the following. If you click on "Details" above next to the failed job, do you see the details?

```
. virtualenv/bin/activate; pylint -j 1 -E --rcfile=./lint-configs/python/.pylintrc --load-plugins=pylint_plugins.api_models --load-plugins=pylint_plugins.db_models st2rbac_backend/
************* Module st2rbac_backend.resolvers
st2rbac_backend/resolvers.py:726:35: E1101: Class 'PermissionType' has no 'KEY_VALUE_LIST' member (no-member)
make: *** [Makefile:115: .pylint] Error 2
```

File: tests/unit/controllers/api/v1/test_kvps_rbac.py [hunk 4]
```diff
@@ -41,360 +47,605 @@ def setUp(self):
...
+        )
+        UserRoleAssignment.add_or_update(role_assignment_db)
+
+        user_db = self.users["user_6"]
+        role_assignment_db = UserRoleAssignmentDB(
+            user=user_db.name,
+            role=self.roles["user_6"].name,
+            source="assignments/%s.yaml" % user_db.name,
+        )
+        UserRoleAssignment.add_or_update(role_assignment_db)
```
Comment (m4dcoder) [inline tests/unit/controllers/api/v1/test_kvps_rbac.py:318]: There are a lot of users, roles, and grants added here. Are these test cases all covered in the unit tests below?  It's hard to follow. Can you add comments what these users/roles/grants are used in which tests?

**Review by m4dcoder — CHANGES_REQUESTED** [review]
Some minor comments that need to be addressed. Otherwise, LGTM.

File: tests/unit/controllers/api/v1/test_kvps_rbac.py [hunk 5]
```diff
@@ -41,360 +47,605 @@ def setUp(self):
...
         self.assertEqual(len(resp.json), self.system_scoped_items_count)
         for item in resp.json:
-            self.assertEqual(item['scope'], FULL_SYSTEM_SCOPE)
+            self.assertEqual(item["scope"], FULL_SYSTEM_SCOPE)
 
         # Verify second item is encrypted
-        self.assertTrue(resp.json[1]['secret'])
-        self.assertTrue(len(resp.json[1]['value']) > 50)
+        self.assertTrue(resp.json[1]["secret"])
+        self.assertTrue(len(resp.json[1]["value"]) > 50)
```
[the remark this answers is not in the document]
Comment (ankita-orchestral) [reply tests/unit/controllers/api/v1/test_kvps_rbac.py:337]: We have formatted the code according to the black tool. This is an already existing code and we haven't done any logical changes to the above code

File: tests/unit/test_rbac_resolvers_key_value_pair.py [hunk 6]
```diff
@@ -17,39 +17,433 @@
...
+
+        user_db = self.users["st2kv_system_grant"]
+        role_assignment_db = UserRoleAssignmentDB(
+            user=user_db.name,
+            role=self.roles["st2kv_system_grant"].name,
+            source="assignments/%s.yaml" % user_db.name,
+        )
+        UserRoleAssignment.add_or_update(role_assignment_db)
+
+    def test_user_resource_db_permissions(self):
```
Comment (m4dcoder) [inline tests/unit/test_rbac_resolvers_key_value_pair.py:300]: This test covers too much. It's also missing some test cases. 

Can you separate this out to multiple tests? Please systematically refactor the unit tests according to the following list.

- System roles
  - System Observer role on all system scoped KVPs
    - Succeed on LIST permission
    - Succeed on VIEW permission
    - Fail on SET permission
    - Fail on DELETE permission
  - System Admin role on all system scoped KVPs
    - Succeed on ALL permission
    
- Default for user on user owned KVPs
  - No specific roles and grants need to be defined for user to set/delete their own KVPs.
  - User should be defined access on ALL permission to another user's KVPs.
  - Succeed on ALL permission to user owned KVPs

- User roles with permission on specific system scope KVPs
  - Custom user role with LIST on specific system scope KVP
    - Succeed on LIST permission
    - Should VIEW permission succeed or fail here?
    - Fail on SET permission
    - Fail on DELETE permission
  - Custom user role with VIEW on specific system scope KVPs
    - Should LIST permission succeed or fail here?
    - Succeed on VIEW permission
    - Fail on SET permission
    - Fail on DELETE permission
  - Custom user role with SET on specific system scope KVPs
    - Should LIST permission succeed or fail here?
    - Succeed on VIEW permission
    - Succeed on SET permission
    - Fail on DELETE permission
  - Custom user role with DELETE on specific system scope KVPs
    - Should LIST permission succeed or fail here?
    - Succeed on VIEW permission
    - Succeed on SET permission
    - Succeed on DELETE permission

File: tests/unit/controllers/api/v1/test_kvps_rbac.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (m4dcoder) [reply tests/unit/controllers/api/v1/test_kvps_rbac.py:337]: We can be good citizen to clean up unit tests. Please review the unit tests in this file and clean up the duplicates. There're only a few.

**Review by m4dcoder — CHANGES_REQUESTED** [review]
Please refactor the unit tests to cover test cases I documented in the comments.

**Comment by ankita-orchestral** [comment]
@m4dcoder - We have refactored the unit test cases as per your comment. Please review and let us know your feedback.

File: tests/unit/test_rbac_resolvers_key_value_pair.py (hunk 6 above)
Comment (m4dcoder) [reply tests/unit/test_rbac_resolvers_key_value_pair.py]: @ashwini-orchestral Please review the outline above. I highlighted the unit tests for each of the test cases in the outline here.

File: tests/unit/test_rbac_resolvers_key_value_pair.py [hunk 7]
```diff
@@ -818,3 +818,86 @@ def test_admin_permissions_for_user_scoped_kvps(self):
             resource_db=kvp_1_db,
             permission_types=self.write_permission_types,
         )
+
+    def test_user_permissions_for_system_scope_kvps(self):
```
Comment (m4dcoder) [inline tests/unit/test_rbac_resolvers_key_value_pair.py:822]: There are already a number of tests on user permissions for system scope kvps rewritten under KeyValueSystemScopePermissionsResolverTestCase in my last commit.  What's added here that's different than those tests?

**Review by m4dcoder — CHANGES_REQUESTED** [review]

File: tests/unit/test_rbac_resolvers_key_value_pair.py (hunk 7 above)
Comment (ankita-orchestral) [reply tests/unit/test_rbac_resolvers_key_value_pair.py]: Added the unit test as the list functionality was pending for user with system scope kvps. It got skipped from our mind that it was already written.
We will take care of this in future commits.

**Comment by amanda11** [comment]
LGTM - with an imminent release of 3.6 do we need to make sure that this only goes in with the paired ST2 pair?

File: st2rbac_backend/resolvers.py [hunk 8]
```diff
@@ -682,13 +684,96 @@ class KeyValuePermissionsResolver(PermissionsResolver):
...
+            permission_types = self.view_grant_permission_types[:] + [permission_type]
+            # Create role, permission grant for the permission types which doesn't exist
+            result = rbac_service.get_role_by_name(name=permission_type)
+            user_role = rbac_service.get_role_assignments_for_user(user_db, include_remote=True)
+            if len(user_role) > 0:
+               role = user_role[0].role
+            if result is None and (role == PermissionType.KEY_VALUE_PAIR_VIEW or role == PermissionType.KEY_VALUE_PAIR_LIST):
+                role_db = rbac_service.create_role(name=permission_type, description="Role of %s" % permission_type)
+                role_details = rbac_service.assign_role_to_user(role_db=role_db, user_db=user_db, source='assignments/%s.yaml' % user_db.name)
+                permission_grant = rbac_service.create_permission_grant_for_resource_db(role_db=role_db, resource_db=resource_db, permission_types=permission_types)
```
Comment (m4dcoder) [inline st2rbac_backend/resolvers.py:759]: Why are we creating permission grant here? Which test case is this supporting. This looks very wrong.

**Review by m4dcoder — CHANGES_REQUESTED** [review]

File: st2rbac_backend/resolvers.py (hunk 8 above)
Comment (ankita-orchestral) [reply st2rbac_backend/resolvers.py]: This was added to handle custom user roles with LIST/SET/DELETE for system scoped kvps.

Comment (m4dcoder) [reply st2rbac_backend/resolvers.py]: Custom user roles for specific system scoped KVPs should be defined explicitly in the YAML files. The resolvers should not be adding any new role assignment and permission grant records into the database automatically.

Comment (ankita-orchestral) [reply st2rbac_backend/resolvers.py]: Yes, that's correct and it works as mentioned.

File: tests/unit/test_rbac_resolvers_key_value_pair.py [hunk 9]
```diff
@@ -15,41 +15,806 @@
...
+
+        # Admin user should have general list permissions on user1's kvps.
+        self.assertUserHasResourceDbPermission(
+            resolver=resolver,
+            user_db=admin_user_db,
+            resource_db=KeyValuePairDB(scope="%s:%s" % (FULL_USER_SCOPE, user_1_db.name)),
+            permission_type=PermissionType.KEY_VALUE_PAIR_LIST,
+        )
+
+        # User2 should have any permissions on another user1's kvp.
```
Comment (cognifloyd) [inline tests/unit/test_rbac_resolvers_key_value_pair.py:800]:
```suggestion
        # Admin should have any permissions on another user1's kvp.
```

**Review by cognifloyd — COMMENTED** [review]
There's a lot here. It will be wonderful to have RBAC on the datastore. :)
I have a couple of questions, but overall I think this is excellent.

**Review by cognifloyd — APPROVED** [review]
Sweet. I don't see anything else that has to be changed before we merge.

File: tests/unit/controllers/api/v1/test_kvps_rbac.py [hunk 10]
```diff
@@ -15,386 +15,569 @@
...
+        resp = self.app.get("/v1/keys?limit=-1", expect_errors=True)
+        self.assertEqual(resp.status_code, http_client.FORBIDDEN)
 
-        resp = self.app.get('/v1/keys?decrypt=True')
+        resp = self.app.get("/v1/keys/")
         self.assertEqual(resp.status_int, 200)
-        self.assertEqual(len(resp.json), self.system_scoped_items_count)
         for item in resp.json:
-            self.assertEqual(item['scope'], FULL_SYSTEM_SCOPE)
+            self.assertEqual(item["scope"], FULL_SYSTEM_SCOPE)
```
[the remark this answers is not in the document]
Comment (m4dcoder) [reply tests/unit/controllers/api/v1/test_kvps_rbac.py]: Refactored the unit tests.

**Review by m4dcoder — APPROVED** [review]
